A107: TLS Private Key Offloading (#524)
grpc/proposal: A107-tls-private-key-offloading.md

using grpc_core::SignatureAlgorithm;
using grpc_core::CustomPrivateKeySignCallback;
using grpc_core::PrivateKey;

// This is the C++ API for PemKeyCertPair
struct IdentityKeyCertPair {
  PrivateKey private_key;
  std::string certificate_chain;
  // Associated ctors, etc
}

class TlsCredentialsOptions {
  // Deprecated. Use `set_identity_certificate_provider` and 
  // `set_root_certificate_provider` instead.
  void set_certificate_provider(
      std::shared_ptr<CertificateProviderInterface> certificate_provider);

  void set_identity_certificate_provider(
      std::shared_ptr<CertificateProviderInterface> certificate_provider);

  void set_root_certificate_provider(
      std::shared_ptr<CertificateProviderInterface> certificate_provider);
}

// Basic wrapping of C-Core
class InMemoryCertificateProvider : public CertificateProviderInterface {

  static std::shared_ptr<InMemoryCertificateProvider> Create(std::string root_cert, std::vector<IdentityKeyCertPair> key_cert_pairs);

  // thread safe updates
  UpdateRoot(std::string root_certificates);
  UpdateIdentity(absl::Span<IdentityKeyCertPair> key_cert_pairs); 
}

// To use, set the provider on the TlsCredentialsOptions
```

> ### Python

Author: Done

      root_certificates: The PEM-encoded root certificates as a byte string,
        or None to retrieve them from a default location chosen by gRPC
        runtime.
      certificate_chain: The PEM-encoded certificate chain as a byte string
        to use or None if no certificate chain should be used.

    Returns:
      A ChannelCredentials for use with an SSL-enabled Channel.
    """

```

We won't significantly refactor the Python API surface \- instead we will allow the `private_key` input to be a signing function.

The most complex piece of this is the implementation \- C-Core/Cython/Python must handle the calling of the user provided Python signing function from C which must invoke a callback that is passed to it. This will involve creating bridge types between the Python user sign function and the expected `absl::AnyInvocable` as well as bridging the callback that is passed to the user sign function while managing the GIL and asynchronous nature of the signing. This is technically feasible with Cython. A proof-of-concept of this structure [is written here.](https://source.corp.google.com/piper///depot/google3/experimental/users/gregorycooke/python_cpp_wrapping/) 

```py

> # Example Usage

Reviewer: Spoke with @matthewstevenson88 - he explained some of the async patterns our users might apply when `CustomPrivateKeySign`. One of them is creating a thread that will do the computation, passing `on_done` callback to the thread, while immediately returning from the `CustomPrivateKeySign`. This is a good argument for not introducing an internal wrapper.

I need think about this some more, and consider a few nuances: 

- Whether a thread will actually be useful for offloading CPU-bound tasks - because if GIL is stuck on this thread waiting for the signing to complete, there won't be any parallelization/concurrency (unless it  calls into a Python C extension that disables GIL)
- Whether we still can use the wrapper approach, but with a different interface, like using the [Future](https://docs.python.org/3/library/concurrent.futures.html#future-objects) interface. `Future` should make it easier for the user to work with various `PoolExecutor`s.
- Whether we should also allow an `async` method signature (make `private_key_sign_fn` a `Union[CustomPrivateKeySign, CustomPrivateKeySignAsync]`) - which may be required if it's common for the HW offloading libraries the do the signing to use async/await pattern ([asyncio](https://docs.python.org/3/library/asyncio.html) python).

> # Example Usage

Reviewer: For the reference, 

> GIL and performance considerations
>
> Unlike the [multiprocessing](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing) module, which uses separate processes to bypass the [global interpreter lock](https://docs.python.org/3/glossary.html#term-global-interpreter-lock) (GIL), the threading module operates within a single process, meaning that all threads share the same memory space. However, the GIL limits the performance gains of threading when it comes to CPU-bound tasks, as only one thread can execute Python bytecode at a time. Despite this, threads remain a useful tool for achieving concurrency in many scenarios.
>
> — https://docs.python.org/3/library/threading.html#gil-and-performance-considerations

# example_signer is a basic implementation CustomPrivateKeySign
def example_signer(
    unsigned_data: bytes,
    algorithm: SignatureAlgorithm,
    on_done: PrivateKeySignDoneCallback,
) -> None:
    # Manually sign the bytes.
    signed_bytes = ...
    if signed_bytes:
        on_done(signed_bytes, True)
    else:
        on_done(None, False)

# Now the user is in their application configuring gRPC 
# Create creds with the custom signer

> creds = ssl_channel_credentials_with_custom_signer(<some_root>, example_signer, <some_chain>)

Reviewer: note - we require arguments to be passed as keyword arguments now. 
```suggestion
creds = ssl_channel_credentials_with_custom_signer(
    private_key_sign_fn=example_signer,
    root_certificates=b"<some_root>",
    certificate_chain=b"<some_chain>",
)
```

```go
// From crypto.tls
type Signer interface {
	// Public returns the public key corresponding to the opaque,
	// private key.
	Public() PublicKey

	// Sign signs digest with the private key, possibly using entropy from
	// rand. For an RSA key, the resulting signature should be either a
	// PKCS #1 v1.5 or PSS signature (as indicated by opts). For an (EC)DSA
	// key, it should be a DER-serialised, ASN.1 signature structure.
	//
	// Hash implements the SignerOpts interface and, in most cases, one can
	// simply pass in the hash function used as opts. Sign may also attempt
	// to type assert opts to other types in order to obtain algorithm
	// specific values. See the documentation in each package for details.
	//
	// Note that when a signature of a hash of a larger message is needed,
	// the caller is responsible for hashing the larger message and passing
	// the hash (as digest) and the hash function (as opts) to Sign.

> 	Sign(rand io.Reader, digest []byte, opts SignerOpts) (signature []byte, err error)

Author: Thanks for the heads up - I took a look at the change and I think someone implementing `crypto.MessageSigner`, and I think it should "just work" since crypto/tls supports a `crypto.MessageSigner` as a `PrivateKey` in the `tls.Certificate`.

So I think gRPC will pretty blindly pass this through while configuring `tls.Config`. If the user implements a `crypto.MessageSigner` with a `SignMessage` it'll use that

```py
# Example Usage

def sync_client_private_key_signer(
    data_to_sign,
    signature_algorithm,
    on_complete,
):
    """
    Takes in data_to_sign and signs it using the test private key with a sync return
    """
    private_key_bytes = client_private_key()
    signature = sign_private_key(
        data_to_sign, private_key_bytes, signature_algorithm
    )
    return signature
```

Concurrent example:

```py
import multiprocessing

def _sign_worker(data_to_sign, signature_algorithm, on_complete):
    """Worker function to perform signing in a separate process."""
    try:
        # In a real use case, this would use a secure method to access the key.
        signature = sign_private_key(
            data_to_sign, client_private_key(), signature_algorithm
        )
        on_complete(signature)
    except Exception as e:
        on_complete(e)

def concurrent_client_private_key_signer(
    data_to_sign, signature_algorithm, on_complete
):
    """Takes in data_to_sign and signs it concurrently using multiprocessing."""
    p = multiprocessing.Process(
        target=_sign_worker,
        args=(data_to_sign, signature_algorithm, on_complete),
    )
    p.start()

>     # Per the Python API, return a callable matching PrivateKeySignCancel

Reviewer: Instead, let's explain the flow? Something like "This will be called when ...`

Author: PTAL

    def cancel():
        p.terminate()
        p.join()

    return cancel

> # In the Python code configuring gRPC, create creds with the custom signer

Reviewer: nit: now this is a part of "Concurrent example", which may be a bit confusing. Let's move it to the top, so the flow is as follows:

---

Usage example:

```py
creds = grpc.experimental.ssl_channel_credentials_with_custom_signer(
    private_key_sign_fn=your_signer_fn
    certificate_chain=your_cert_chain,
    root_certificates=your_root_certs,
)
```

Synchronous (blocking) signer example:

```py
def sync_client_private_key_signer(
    data_to_sign,
    signature_algorithm,
    on_complete,
): ...
```

Concurrent signer example:

```py
import multiprocessing
...
```

Author: Good call, moved around

function should expect a *digest* of the transcript as input.

```go
// From crypto.tls
type Signer interface {
	// Public returns the public key corresponding to the opaque,
	// private key.
	Public() PublicKey

	// Sign signs digest with the private key, possibly using entropy from
	// rand. For an RSA key, the resulting signature should be either a
	// PKCS #1 v1.5 or PSS signature (as indicated by opts). For an (EC)DSA
	// key, it should be a DER-serialised, ASN.1 signature structure.
	//
	// Hash implements the SignerOpts interface and, in most cases, one can
	// simply pass in the hash function used as opts. Sign may also attempt
	// to type assert opts to other types in order to obtain algorithm
	// specific values. See the documentation in each package for details.
	//
	// Note that when a signature of a hash of a larger message is needed,
	// the caller is responsible for hashing the larger message and passing
	// the hash (as digest) and the hash function (as opts) to Sign.
	Sign(rand io.Reader, digest []byte, opts SignerOpts) (signature []byte, err error)
}

cert = tls.Certificate{
Certificate: derChain, //[][]byte

> PrivateKey:  PrivateKeySigner, //crypto.PrivateKey

Reviewer 2: Nit: Can we add a tab to these fields in this literal struct initialization syntax to make it more readable. Thanks.

Author: Done

This will be piped around by the existing infrastructure to configure the
`crypto/tls` library with no additional changes needed.

```go
// Example Usage

// ExampleSigner is a basic implementation of crypto.Signer
type ExampleSigner struct {}

// Public returns the public key corresponding to the opaque, private key.
func (s *ExampleSigner) Public() crypto.PublicKey {
	return <some key>
}

// Sign signs the digest with the private key.
func (s *ExampleSigner) Sign(r io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
       signed_bytes := <do stuff with digest>
	return signed_bytes, nil
}

// Now the user is in their application configuring gRPC 
// Create the custom signer  
customSigner, err := NewExampleSigner()

// Create the tls.Certificate
tlsCert := &tls.Certificate{
    Certificate: <pem-certificate>
    Signer: customSigner
}

// Create the InMemoryCertProvider and configure gRPC with it
provider, err := NewInMemoryCertProvider(tlsCert)
serverOpts := &advancedtls.Options{
	IdentityOptions: advancedtls.IdentityCertificateOptions{
		IdentityProvider: provider,
	},
}

// Down the line update
provider.UpdateCertificate(&tls.Certificate(<some updated thing>))

> 

Reviewer 2: For completeness sake, does it also make sense to show how this options struct will be used to create a credentials and how it will be passed to create the grpc channel?

Author: Yes - I actually removed this line about UpdateCertificate because I think it's just confusing for this example.

A107 - TLS Private Key Offloading
----
* Author: @gtcooke94
* Approver: ejona86

> * Status: C++ implemented

Reviewer 3: This was the status of the gRFC, with a fixed enumeration of values. So it would have just been "Implemented". The next line would give specifics as to the languages that are complete.

Although with #275 just merged, you could delete this line instead. We removed it because it was essentially always wrong and multiple of the states only applied while it was a PR which didn't provide value.

Author: Deleted

> * Implemented in: C++, Go

Reviewer 3: This is for languages that already have it implemented. Is Go already implemented? I don't see `InMemoryCertProvider`; was it renamed?

Author: Just C++, removed Go

```

This will be piped around by the existing infrastructure to configure the
`crypto/tls` library with no additional changes needed.

```go
// Example Usage

// ExampleSigner is a basic implementation of crypto.Signer
type ExampleSigner struct {}

// Public returns the public key corresponding to the opaque, private key.
func (s *ExampleSigner) Public() crypto.PublicKey {
	return <some key>
}

// Sign signs the digest with the private key.
func (s *ExampleSigner) Sign(r io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
       signed_bytes := <do stuff with digest>
	return signed_bytes, nil
}

// Now the user is in their application configuring gRPC 
// Create the custom signer  
customSigner, err := NewExampleSigner()

// Create the tls.Certificate
tlsCert := &tls.Certificate{
    Certificate: <pem-certificate>
    Signer: customSigner
}

// Create the InMemoryCertProvider and configure gRPC with it
provider, err := NewInMemoryCertProvider(tlsCert)
clientOpts := &advancedtls.Options{
	IdentityOptions: advancedtls.IdentityCertificateOptions{
		IdentityProvider: provider,
	},
}

> clientTLSCreds, err := advancedtls.NewClientCreds(options)

Reviewer 3: s/options/clientOpts/ ?

Author: Fixed

```

### Java

After initial investigation, we will not pursue implementing this feature in
Java right now. Due to the different designs of the options of underlying
security library APIs, there is no cohesive convenience API that we could add to
gRPC-Java. A user can still implement this themselves and use the
`AdvancedTlsKeyProvider` APIs to do whatever they want with the `PrivateKey`
interface. Particularly, a user could create their own signature provider and
globally register it to be used. gRPC-Java, as a library, will not do global
registration.

 

### Temporary environment variable protection

This feature will be explicitly configured by users, thus no environment
variable protection is needed. If a user does not configure TLS Private Key
Offloading, it will not happen.

## Rationale

Private key offloading is designed to support signing outside of the existing
process, for example in a hardware module or via an RPC \- thus this API should
support asynchronous operations in languages where that is possible (Golang's
crypto/tls does **not** support asynchronous private key signing). 

We are largely restricted by the underlying security libraries in each language.

> In the following sections, each language's API will be discussed as they are

Reviewer 3: "following sections" accidentally left over after this text moved?

Author: Moved it above each language section

interface-based approach from C++.  It returns either how to cancel the async
operation in the async case or a signature in the sync case.

Note: gevent is NOT supported.

```py

# A Callable to return in the async case
# See the `ssl_channel_credentials_with_custom_signer` docstring for more detail on usage.
PrivateKeySignCancel = Callable[[], None]
PrivateKeySignatureAlgorithm = _cygrpc.PrivateKeySignatureAlgorithm
PrivateKeySignOnComplete = Callable[[Union[bytes, Exception]], None]

# See the `ssl_channel_credentials_with_custom_signer` docstring for more detail on usage.
# The custom signing function for a user to implement and pass to gRPC Python.
CustomPrivateKeySign = Callable[
    [
        bytes,
        PrivateKeySignatureAlgorithm,
        "PrivateKeySignOnComplete",
    ],
    Union[bytes, "PrivateKeySignCancel"],
]

@experimental_api
def ssl_channel_credentials_with_custom_signer(
    *,
    private_key_sign_fn: "CustomPrivateKeySign",
    root_certificates: Optional[bytes] = None,
    certificate_chain: bytes,
) -> grpc.ChannelCredentials:
    """Creates a ChannelCredentials for use with an SSL-enabled Channel with a custom signer.

    THIS IS AN EXPERIMENTAL API.

>     This API will be removed in a future version and combined with `grpc.ssl_channel_credentials`.

Reviewer: Let's add the note on this not working with gevent here too.

Author: Done

        or None to retrieve them from a default location chosen by gRPC
        runtime.
      certificate_chain: The PEM-encoded certificate chain as a byte string
        to use

    Returns:
      A ChannelCredentials for use with an SSL-enabled Channel.
    """

```

The most complex piece of this is the implementation \- C-Core/Cython/Python
must handle the calling of the user provided Python signing function from C
which must invoke a callback that is passed to it. This will involve creating
bridge types between the Python user sign function and the expected
`absl::AnyInvocable` as well as bridging the callback that is passed to the user
sign function while managing the GIL and asynchronous nature of the signing.
This is technically feasible with Cython.

```py
# Example Usage

def sync_client_private_key_signer(
    data_to_sign,
    signature_algorithm,
    on_complete,
):
    """
    Takes in data_to_sign and signs it using the test private key with a sync return
    """
    private_key_bytes = client_private_key()
    signature = sign_private_key(
        data_to_sign, private_key_bytes, signature_algorithm
    )
    return signature

> 

Reviewer: Let's separate these two examples, make it

```md
(break triple tick)

Concurrent example:

(new triple tick)
```

Author: Done

> def async_signer_worker(data_to_sign, signature_algorithm, on_complete):

Reviewer: ```suggestion
def _thread_signer_worker(data_to_sign, signature_algorithm, on_complete):
```

Author: Done

Concurrent example:

> ```

Reviewer: nit: add `py` syntax

Author: Done

>     Meant to be used as an async function for a thread, for example

Reviewer: Let's make it clear it's used in `concurrent_client_private_key_signer`

Author: Done

>     """

Reviewer: Async function has a special meaning in python: https://docs.python.org/3/library/asyncio.html. I recommend changing the term to "concurrent":
```suggestion
def concurrent_signer_worker(data_to_sign, signature_algorithm, on_complete):
    """
    Meant to be used as an concurrent function for a thread, for example
    """
```

Author: Done

    signature = sign_private_key(
        data_to_sign, private_key_bytes, signature_algorithm
    )
    on_complete(signature)

> def no_op_cancel():

Reviewer: ```suggestion
def _no_op_cancel():
```

Author: Done

>     pass

Reviewer: Let's add a comment explaining why we pass:

> Threads cannot be destroyed, stopped, suspended, resumed, or interrupted
> https://docs.python.org/3/library/threading.html#threading.TIMEOUT_MAX

We should clarify that in other cases (like sending external RPC, or running a separate process) the logic should be to cancel the underlying concurrent call.

Reviewer: Also, let's make it clear it's used in `concurrent_client_private_key_signer`

Reviewer 4: This comment made me realize that this is actually a very bad example to use here, because we really don't want users writing key signing implementations that don't support cancellation.

Please replace this example with one that can actually cancel the operation.  For example, consider a case where the key signer might issue an RPC to do the key signing; when the signing request gets cancelled, the underlying RPC can also be cancelled.

Author: Swapped to a more pseduo-code example

Reviewer: I agree with Mark - having an actual cancellable example is much better than using `def on_cancel: pass` antipattern. 

However, I've looked at the new pseudo-code example, and not sure if it's useful anymore. At least the thread demonstrated the flow better. Now it's not clear where "on_complete" is called and why `concurrent_client_private_key_signer` needs `_do_signing_concurrent` in the first place.

I think we should replace it with more realistic cancellable example, f.e. using something like `multiprocessing.Process` or RPC cancellation https://github.com/grpc/grpc/tree/master/examples/python/cancellation#cancellation-on-the-client-side.

@gtcooke94 — Let @asheshvidyut know if you need help drafting that.

Author: Done, PTAL

Reviewer: Looks so much better, thank you!

> def async_client_private_key_signer(

Reviewer: Async function has a special meaning in python: https://docs.python.org/3/library/asyncio.html. I recommend changing the term to "concurrent":

```suggestion
def concurrent_client_private_key_signer(
```

Author: Done

    data_to_sign, signature_algorithm, on_complete
):
    """
    Takes in data_to_sign and signs it using the test private key
    """
    threading.Thread(
        target=async_signer_worker,
        args=(data_to_sign, signature_algorithm, on_complete),
    ).start()
    return no_op_cancel

# Now the user is in their application configuring gRPC 

> # Create creds with the custom signer

Reviewer: reword?

Author: Done

> creds = ssl_channel_credentials_with_custom_signer(<some_root>, <sync or async example_signer, <some_chain>)

Reviewer: Note that we we require kwargs now.

```suggestion
creds = grpc.experimental.ssl_channel_credentials_with_custom_signer(
    private_key_sign_fn=your_signer_fn
    certificate_chain=your_cert_chain,
    root_certificates=your_root_certs,
)
```

Author: Done

```

 

### Temporary environment variable protection

This feature will be explicitly configured by users, thus no environment
variable protection is needed. If a user does not configure TLS Private Key
Offloading, it will not happen.

## Rationale

Private key offloading is designed to support signing outside of the existing
process, for example in a hardware module or via an RPC \- thus this API should
support asynchronous operations in languages where that is possible (Golang's
crypto/tls does **not** support asynchronous private key signing). 

We are largely restricted by the underlying security libraries in each language.
In the following sections, each language's API will be discussed as they are
dependent upon the SSL library interfaces. Further, each language has different
expectations for the sign functions on whether raw bytes or a digest is
expected.

## Implementation

### C-Core

#### *APIs*

BoringSSL provides an asynchronous API for private key signing, so we will
provide an asynchronous, cancellable API using callbacks. The C-Core APIs used
to configure this feature will rely on a new certificate provider
implementation.

> We will create an `InMemoryCertificateProvider` that takes a set of root certs

Reviewer 4: Before the subsections describing each of the three changes in detail, please add a bulletted list of the three changes, with just a few words about each to say why we're doing it.  That way, the reader has a bit of context about why we're making these changes before we dive into the details.

> We will create an `InMemoryCertificateProvider` that takes a set of root certs

Author: Done

	return signed_bytes, nil
}

// Now the user is in their application configuring gRPC 
// Create the custom signer  
customSigner, err := NewExampleSigner()

// Create the tls.Certificate
tlsCert := &tls.Certificate{
    Certificate: <pem-certificate>
    Signer: customSigner
}

// Create the InMemoryCertProvider and configure gRPC with it
provider, err := NewInMemoryCertProvider(tlsCert)
serverOpts := &advancedtls.Options{
	IdentityOptions: advancedtls.IdentityCertificateOptions{
		IdentityProvider: provider,
	},
}

// Down the line update
provider.UpdateCertificate(&tls.Certificate(<some updated thing>))

```

### Java

After initial investigation, we will not pursue implementing this feature in
Java right now. Due to the different designs of the options of underlying
security library APIs, there is no cohesive convenience API that we could add to
gRPC-Java. A user can still implement this themselves and use the
`AdvancedTlsKeyProvider` APIs to do whatever they want with the `PrivateKey`
interface. Particularly, a user could create their own signature provider and
globally register it to be used. gRPC-Java, as a library, will not do global
registration.

# **Observability**

> The following non-per-call metric and labels will be added. These will allow a

Author: Done

user insight into the offloaded operations and will be an aid in debugging
failures.

| Name | Disposition | Description |
| :---- | :---- | :---- |
| `grpc.target` | required | Indicates the target of the gRPC channel for which this handshake is occurring |
| `grpc.security.handshaker.offloaded_private_key_sign.algorithm` | optional | The signature algorithm used to sign |
| `grpc.security.handshaker.offloaded_private_key_sign.status` | optional | The status code return for the private key sign function |

| Metric | Type | Unit | Labels | Description |
| :---- | :---- | :---- | :---- | :---- |

> | `grpc.security.handshaker.offloaded_private_key_sign_duration` | Histogram | float64/double s | `grpc.target, grpc.security.handshaker.offloaded_private_key_sign.algorithm, grpc.security.handshaker.offloaded_private_key_sign.status` | How long the offloaded private key signing took |

Reviewer 4: Still need to resolve this question.

### Related Proposals: 
* [A66]

> * [A79]

Author: Done

```

For TLS\>=1.2 with ECDHE, a requirement for this feature, the `decrypt` method is not needed. We will provide an
interface through which users can provide a custom private key signer, with the
higher-level functionality being a function that takes in unsigned bytes and a
signature algorithm and returns signed bytes.

```
string signed_bytes sign(string algorithm, string unsigned_bytes)
```

> ### C-Core

Author: Done

#### InMemory Certificate Provider
We will create an `InMemoryCertificateProvider` that takes a set of root certs
and a list of `PemKeyCertPair`. Further, these values can be manually updated by
the user. This provides a strict super-set of functionality of the current
`StaticDataCertificateProvider` - calling `UpdateRoot` and `UpdateIdentity` on
an `InMemoryCertificateProvider` at setup is the same as a

> `StaticDataCertificateProvider`.

Author: Done

The updates will return an `absl::Status` to the caller, and further a
`ValidateCredentials` API is provided.

> ```c

Author: Done

// Implements a provider that uses in-memory data that can be modified in a thread-safe manner.
class InMemoryCertificateProvider : grpc_tls_certificate_provider {
 public:
  InMemoryCertificateProvider(std::string root_certificates, PemKeyCertPairList pem_key_cert_pairs);
  
  // Thread safe updates
  absl::Status UpdateRoot(std::string root_certificates);
  absl::Status UpdateIdentityKeyCertPair(
      const std::vector<IdentityKeyCertPair>& identity_key_cert_pairs);

  // Returns an OK status if the following conditions hold:
  // - the root certificates consist of one or more valid PEM blocks, and
  // - every identity key-cert pair has a certificate chain that consists of
  //   chain that consists of valid PEM blocks and has a private key is a valid
  //   PEM block.
  absl::Status ValidateCredentials() const;
}
```

#### Splitting Certificate Providers into Identity and Root Providers
We will also decouple the root provider and the identity provider in [the

> tls\_credentials\_options](http://google3/third_party/grpc/src/core/credentials/transport/tls/grpc_tls_credentials_options.h;l=77;rcl=731487339)

Author: Oops, I keep thinking I've gotten them all

and through the stack.  Users should be able to specify, for example, an
`InMemoryCertificateProvider` that implements private key offloading and a
`FileWatcherCertificateProvider` for the root certificates. We will mark the
[coupled
API](http://google3/third_party/grpc/include/grpcpp/security/tls_credentials_options.h;l=55-56;rcl=682352913)
as deprecated.  This is required for this effort, as specifying a private key
signer must be done for the identity, but is unrelated to how the root is
provided. Currently, identity (private key and certificate chain) presentation
to gRPC is coupled with the roots presentation to gRPC.

```c++
class TlsCredentialsOptions {
 public:

>   [[deprecated(

Author: Done

      "Use set_root_certificate_provider() or "
      "set_identity_certificate_provider() instead.")]]
  void set_certificate_provider(
      std::shared_ptr<CertificateProviderInterface> certificate_provider);
  void set_root_certificate_provider(
      std::shared_ptr<CertificateProviderInterface> certificate_provider);
  void set_identity_certificate_provider(
      std::shared_ptr<CertificateProviderInterface> certificate_provider);
```

> #### Private Key Signing

Author: Done

> BoringSSL provides an asynchronous API for private key signing, so we will

Author: Done

provide an asynchronous, cancellable API using callbacks. The C-Core APIs used
to configure this feature will rely on a new certificate provider
implementation.

The implementation of `PrivateKeySigner::Sign` can be sync or async. In the sync
case, the implementer must return a signature or a non-ok status.  In the async
case, the implementer must kick off their async operations then return a
`PrivateKeySigner::AsyncSigningHandle`. The async operation must call the
provided `on_sign_complete` function with the signing result when the operation
is complete. The implementer must also implement `PrivateKeySigner::Cancel`.
`Cancel` will be called by gRPC in the case that an in-flight operation must be
shutdown, and the implementer should use this function to ensure any resources
being used for an asynchronous signing operation are properly shut down and

> released.

Author: Took out most of the detail, left the bit about the user being responsible for cleaning up resources during Cancellation

```c++

> using PrivateKey = std::variant<absl::string_view, CustomPrivateKeySign>

Author: Removed

// This already exists - it will be deprecated and replaced with `IdentityKeyOrSignerCertPair` where necessary.
struct [[deprecated("Use IdentityKeyOrSignerCertPair instead")]] GRPCXX_DLL
    IdentityKeyCertPair {
  std::string private_key;
  std::string certificate_chain;
};

// A struct that stores the credential data presented to the peer in handshake
// to show local identity. The private_key and certificate_chain should always
// match. The private_key can be either a PEM string or a PrivateKeySigner.
// The PrivateKeySigner will only work with gRPC binaries compiled with
// BoringSSL.

> struct GRPCXX_DLL IdentityKeyOrSignerCertPair {

Author: Added

```c++
class TlsCredentialsOptions {
 public:
  // The existing method.
  [[deprecated(
      "Use set_root_certificate_provider() or "
      "set_identity_certificate_provider() instead.")]]
  void set_certificate_provider(
      std::shared_ptr<CertificateProviderInterface> certificate_provider);
  // The new method proposed in this gRFC.
  void set_root_certificate_provider(
      std::shared_ptr<CertificateProviderInterface> certificate_provider);
  // The new method proposed in this gRFC.
  void set_identity_certificate_provider(
      std::shared_ptr<CertificateProviderInterface> certificate_provider);
```

#### Private Key Signing

BoringSSL provides an asynchronous API for private key signing, so we will
provide an asynchronous, cancellable API using callbacks. Thus, this is only supported in gRPC builds with BoringSSL. If a
`PrivateKeySigner` is used in a non-BoringSSL build, the user should expect

> failure.

Reviewer 4: Let's be more specific here.  What exactly will fail?  I assume the failure won't be seen until handshake time, right?

Author: Update - it should fail when we try to configure it (the update fn should return a bad status) - https://github.com/grpc/grpc/blob/fb1b22f38a5232744439e3c290913f50da3f927c/src/core/credentials/transport/tls/grpc_tls_certificate_distributor.cc#L342-L345

This flow happens when the user calls `UpdateIdentityKeyCertPair` on their `InMemoryCertificateProvider`.

  // from a previous call to Sign().
  virtual void Cancel(std::shared_ptr<AsyncSigningHandle> handle) = 0;
};
```

The C-Core APIs used
to configure this feature will rely on a new certificate provider

> implementation. 

Reviewer 4: We've already described the `InMemoryCertificateProvider` above, so we don't need to introduce the fact that we're adding a new cert provider impl here.  Instead, this should just say something like "The private key signer can be configured via the `InMemoryCertificateProvider` using a new `IdentityKeyOrSignerCertPair` struct, as shown below."

Author: Done

};
```

The private key signer can be configured via the `InMemoryCertificateProvider`

> using a new `IdentityKeyOrSignerCertPair` struct, as shown below."

Reviewer 4: Please remove trailing `"`.

Author: Done

variable protection is needed. If a user does not configure TLS Private Key
Offloading, it will not happen.

## Rationale

Private key offloading is designed to support signing outside of the existing
process, for example in a hardware module or via an RPC \- thus this API should
support asynchronous operations in languages where that is possible (Golang's
crypto/tls does **not** support asynchronous private key signing). 

We are largely restricted by the underlying security libraries in each language.
In the following sections, each language's API will be discussed as they are
dependent upon the SSL library interfaces. Further, each language has different
expectations for the sign functions on whether raw bytes or a digest is
expected.

## Implementation

* https://github.com/grpc/grpc/pull/40878 - Migrate Python to TlsCredentials under the hood
* https://github.com/grpc/grpc/pull/41490 - Separate cert provider into a root and identity provider
* https://github.com/grpc/grpc/pull/41484 - Create InMemoryCertificateProvider
* https://github.com/grpc/grpc/pull/41606 - Implement PrivateKeySigner in C-Core and C++

> * https://github.com/grpc/grpc/pull/41701 - Implement in Python and Cython

Author: Done

# **Observability**

> The following non-per-call metrics [A79] and labels will be added. These will

Author: Done

allow a user insight into the offloaded operations and will be an aid in
debugging failures.

> | Label Name | Disposition | Description |

Author: Done

| :---- | :---- | :---- |

> | `grpc.target` | required | Indicates the target of the gRPC channel for which this handshake is occurring |

Author: Done

| `grpc.tls.private_key_sign_algorithm` | optional | The signature algorithm used to sign |
| `grpc.status` | optional | The status code return for the private key sign function. From [A66] |

The `grpc.tls.private_key_sign_algorithm` label will contain both a name and key

> length. For example, `RsaPkcs1Sha256`. For more, see the `SignatureAlgorithm`

Reviewer 4: We will want the values for this label to be the same in all languages.  Is that going to be a problem if we define the values via a C++ enum?

Author: I think there are relatively standard names - we'll want to convert to a string and not use the enum directly. Called that out

> enum in the C section above.

Author: Done

> | Metric | Type | Unit | Labels | Description |

Author: Done

## Implementation

* C-Core/C++

>     * https://github.com/grpc/grpc/pull/40878 - Migrate Python to TlsCredentials under the hood

Reviewer 4: This one belongs in the python list, not the C-core/C++ list.
